sky/serve/autoscalers.py

```diff
@@ -222,24 +262,35 @@ def _get_replica_ids_to_scale_down(num_limit: int) -> List[int]:
 
             status_order = serve_state.ReplicaStatus.scale_down_decision_order()
             launched_replica_infos_sorted = sorted(
-                launched_replica_infos,
+                launched_new_replica_infos,
                 key=lambda info: status_order.index(info.status)
                 if info.status in status_order else len(status_order))
 
             return [info.replica_id for info in launched_replica_infos_sorted
                    ][:num_limit]
 
-        if num_launched_replicas < self.target_num_replicas:
+        # Case 1. Once there is min_replicas number of
+        # ready new replicas, we will direct all traffic to them,
+        # we can scale down all old replicas.
+        if len(ready_new_replica_infos) >= self.min_replicas:
+            for info in old_replicas_infos:
+                all_replica_ids_to_scale_down.append(info.replica_id)
+
+        # Case 2. when launched_new_replica_infos is less
+        # than target_num_replicas, we always scale up new replicas.
+        if len(launched_new_replica_infos) < self.target_num_replicas:
             num_replicas_to_scale_up = (self.target_num_replicas -
-                                        num_launched_replicas)
+                                        len(launched_new_replica_infos))
```

Comment:
Just curious, if we only use the number of launched replicas, how do we make sure those replicas being launched are counted? Will `num_replicas_to_scale_up` larger than the number that is actually needed, as there are some replicas already being launched?

Reply:
The number of launched replicas will include replicas that are still launching, in addition to replicas that are ready.

Reply:
I see, it might be better to update the variable name to reflect that, e.g., `provisioning_and_launched_replicas`.